File: packages/assistant-client-ts/src/core/sse.ts
export const DONE_PAYLOAD = "[DONE]";
export const KEEPALIVE_FRAME = ": keep-alive\n\n";

const FRAME_TERMINATOR = "\n\n";
const ID_FIELD = "id: ";
const DATA_FIELD = "data: ";

/** A frame the wire protocol does not define. */
export class MalformedFrameError extends Error {
  readonly raw: string;

  constructor(raw: string) {
    super(`frame is not id/data/comment: ${JSON.stringify(raw)}`);
    this.name = "MalformedFrameError";
    this.raw = raw;
  }
}

/** One SSE frame: its cursor, its payload, and the bytes it arrived as. */
export interface Frame {
  readonly raw: string;
  readonly eventId?: number;
  readonly data?: string;
}

export function isComment(frame: Frame): boolean {
  return frame.eventId === undefined && frame.data === undefined;
}

export function isDone(frame: Frame): boolean {
  return frame.data === DONE_PAYLOAD;
}

export function frameText(eventId: number, data: string): string {
  return `id: ${String(eventId)}\ndata: ${data}${FRAME_TERMINATOR}`;
}

function isDecimal(value: string): boolean {
  return value.length > 0 && /^[0-9]+$/.test(value);
}

/** Read one frame. Throws when it is neither an event nor a comment. */
export function parseFrame(raw: string): Frame {
  if (!raw.endsWith(FRAME_TERMINATOR)) throw new MalformedFrameError(raw);
  const lines = raw.slice(0, -FRAME_TERMINATOR.length).split("\n");
  if (lines.every((line) => line.startsWith(":"))) return { raw };
  let eventId: number | undefined;
  let data: string | undefined;
  for (const line of lines) {
    if (line.startsWith(ID_FIELD)) {
      const cursor = line.slice(ID_FIELD.length);
      if (!isDecimal(cursor)) throw new MalformedFrameError(raw);
      eventId = Number.parseInt(cursor, 10);
    } else if (line.startsWith(DATA_FIELD)) {
      data = line.slice(DATA_FIELD.length);
    } else {
      throw new MalformedFrameError(raw);
    }
  }
  if (eventId === undefined || data === undefined) throw new MalformedFrameError(raw);
  return { raw, eventId, data };
}

export interface ReadFramesOptions {
  /**
   * Treat bytes with no terminator at end of stream as a closed connection
   * rather than as a protocol break.
   */
  allowTruncatedTail?: boolean;
}

/** Split a byte stream into frames and parse each one strictly. */
export async function* readFrames(
  stream: ReadableStream<Uint8Array>,
  options: ReadFramesOptions = {},
): AsyncGenerator<Frame> {
  const reader = stream.getReader();
  const decoder = new TextDecoder();
  let buffer = "";
  try {
    for (;;) {
      const next = await reader.read();
      if (next.done === true) break;
      buffer += decoder.decode(next.value, { stream: true });
      let boundary = buffer.indexOf(FRAME_TERMINATOR);
      while (boundary >= 0) {
        const raw = buffer.slice(0, boundary + FRAME_TERMINATOR.length);
        buffer = buffer.slice(boundary + FRAME_TERMINATOR.length);
        yield parseFrame(raw);
        boundary = buffer.indexOf(FRAME_TERMINATOR);
      }
    }
    buffer += decoder.decode();
    if (buffer.length > 0 && options.allowTruncatedTail !== true) {
      throw new MalformedFrameError(buffer);
    }
  } finally {
    reader.releaseLock();
  }
}

File: packages/assistant-core/tests/sse.py
"""A strict SSE reader, written from the protocol and not from the emitter.

The suite parses frames the way a non-JS consumer must: split on the blank
line, read ``id:`` and ``data:`` fields, treat a leading colon as a comment.
A frame the runtime emits that this reader cannot name is a protocol break.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any
from uuid import UUID

from assistant_core.conversation.event_stream import iter_sse

DONE_PAYLOAD = "[DONE]"
KEEPALIVE_FRAME = ": keep-alive\n\n"


class MalformedFrameError(AssertionError):
    """A frame the wire protocol does not define."""

    def __init__(self, raw: str) -> None:
        super().__init__(f"frame is not id/data/comment: {raw!r}")
        self.raw = raw


@dataclass(frozen=True)
class Frame:
    """One SSE frame: its cursor, its payload, and the bytes it arrived as."""

    raw: str
    event_id: int | None = None
    data: str | None = None

    @property
    def is_comment(self) -> bool:
        return self.event_id is None and self.data is None

    @property
    def is_done(self) -> bool:
        return self.data == DONE_PAYLOAD

    def chunk(self) -> dict[str, Any]:
        assert self.data is not None
        parsed: dict[str, Any] = json.loads(self.data)
        return parsed


def parse_frame(raw: str) -> Frame:
    """Read one frame. Raises when it is neither an event nor a comment."""
    if not raw.endswith("\n\n"):
        raise MalformedFrameError(raw)
    lines = raw[:-2].split("\n")
    if all(line.startswith(":") for line in lines):
        return Frame(raw=raw)
    event_id: int | None = None
    data: str | None = None
    for line in lines:
        if line.startswith("id: "):
            cursor = line.removeprefix("id: ")
            if not cursor.isdigit():
                raise MalformedFrameError(raw)
            event_id = int(cursor)
        elif line.startswith("data: "):
            data = line.removeprefix("data: ")
        else:
            raise MalformedFrameError(raw)
    if event_id is None or data is None:
        raise MalformedFrameError(raw)
    return Frame(raw=raw, event_id=event_id, data=data)


async def read_stream(conversation_id: UUID, *, after: int = 0) -> list[Frame]:
    """Every frame from ``after`` up to and including the turn terminator."""
    return [
        parse_frame(raw)
        async for raw in iter_sse(conversation_id=conversation_id, after=after)
    ]
